### DISCUSSION TITLE
Parsing of Snippets / with subrule

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1404] [Q&A]
Hi, is there a way to parse a string into a ast/document with a subrule?
constructing a on the fly grammar with.
createServicesForGrammar seems to work but also to be tedious when you need to merge with the original grammar

**Comment by spoenemann** [comment]
Currently the only way is to create a clone of the grammar, change the entry rule, and derive a new LangiumServices from that.

We could look into adding an optional parameter to the parser's `parse` method so you can change the main rule. But I don't know how feasible that is. For sure it would require to pass the `rules` map from the parser builder (`createParser` function) to the LangiumParser.

**Reply by cdietrich** [reply]
can i programmatically clone the grammar? (with document) so i could pass the grammar object instead of grammar text to the createServicesForGrammar

**Reply by cdietrich** [reply]
rn i do something like

```
const originalGrammar = XxxxxGrammar()
// create a new grammar with different entry rule
const { shared, grammar: grammarServices } = createLangiumGrammarServices(EmptyFileSystem)
const linker = grammarServices.references.Linker
const buildReference = linker.buildReference.bind(linker)
const copiedGrammar: Grammar = AstUtils.copyAstNode(originalGrammar, buildReference)
for (const rule of copiedGrammar.rules) {
  if (GrammarAST.isParserRule(rule)) {
    if (rule.name === "Xxxxxx") {
      rule.entry = false
    }
    if (rule.name === "Expression") {
      rule.entry = true
    }
  }
}

const doc = shared.workspace.LangiumDocumentFactory.fromModel(copiedGrammar, URI.parse("dummy.langium"))

services = await createServicesForGrammar({
  grammar: doc.parseResult.value,
  grammarServices: grammarServices,
})
```

**Reply by spoenemann** [reply]
I made an experiment in #1407 to simplify this. Could you check whether it works for you?

**Reply by cdietrich** [reply]
need to figure out how to consume this in our env.

**Reply by cdietrich** [reply]
cool it seems to work. can we expose this on `parseHelper` via flag in `ParseHelperOptions`?
